File: Days 71 - 80/Day 73/ReverseLinkedList.cpp
#include <iostream>
#include <memory>
#include <string>

struct Node
{
	std::string value;

	std::shared_ptr<Node> next;

	explicit Node(const std::string& value, std::shared_ptr<Node> next = nullptr) noexcept
		: value(value), next(next)
	{ }
};

std::shared_ptr<Node> ReverseLinkedListHelper(std::shared_ptr<Node> head, std::shared_ptr<Node> next) noexcept
{
	if (head == nullptr)
	{
		return next;
	}

	std::shared_ptr<Node> previous = head->next;
	head->next = next;

	return ReverseLinkedListHelper(previous, head);
}

void ReverseLinkedList(std::shared_ptr<Node> head) noexcept
{
	ReverseLinkedListHelper(head, nullptr);
}

void PrintLinkedList(std::shared_ptr<Node> head) noexcept
{
	while (head != nullptr)
	{
		std::cout << head->value << "\n";
		head = head->next;
	}
}

int main(int argc, char* argv[])
{
	std::shared_ptr<Node> n1 = std::make_shared<Node>("1");
	std::shared_ptr<Node> n2 = std::make_shared<Node>("2", n1);
	std::shared_ptr<Node> n3 = std::make_shared<Node>("3", n2);
	std::shared_ptr<Node> n4 = std::make_shared<Node>("4", n3);
	std::shared_ptr<Node> n5 = std::make_shared<Node>("5", n4);
	std::shared_ptr<Node> n6 = std::make_shared<Node>("6", n5);
	std::shared_ptr<Node> n7 = std::make_shared<Node>("7", n6);
	std::shared_ptr<Node> n8 = std::make_shared<Node>("8", n7);
	
	std::cout << "Printing from tail:\n";
	PrintLinkedList(n1);

	std::cout << "\nPrinting from head:\n";
	PrintLinkedList(n8);

	std::cout << "\nREVERSING LIST...\n\n";
	ReverseLinkedList(n8);

	std::cout << "Printing from old tail:\n";
	PrintLinkedList(n1);

	std::cout << "\nPrinting from old head:\n";
	PrintLinkedList(n8);

	std::cin.get();

	return 0;
}

File: Days 071 - 080/Day 73/ReverseLinkedList.cs
using System;

namespace DailyCodingProblem
{
	internal class Node
	{
		public string Value { get; set; }
		public Node Next { get; set; }

		public Node(string value, Node next = null)
		{
			this.Value = value;
			this.Next = next;
		}
	}

	internal class Day73
	{
		private static int Main(string[] args)
		{
			Node n1 = new Node("1");
			Node n2 = new Node("2", n1);
			Node n3 = new Node("3", n2);
			Node n4 = new Node("4", n3);
			Node n5 = new Node("5", n4);
			Node n6 = new Node("6", n5);
			Node n7 = new Node("7", n6);
			Node n8 = new Node("8", n7);

			Console.WriteLine("Printing from tail:");
			PrintLinkedList(n1);

			Console.WriteLine("\nPrinting from head:");
			PrintLinkedList(n8);

			Console.WriteLine("\nREVERSING LIST...\n");
			ReverseLinkedList(n8);

			Console.WriteLine("Printing from old tail:");
			PrintLinkedList(n1);

			Console.WriteLine("\nPrinting from old head:");
			PrintLinkedList(n8);

			Console.ReadLine();

			return 0;
		}

		private static void ReverseLinkedList(Node head) => ReverseLinkedListHelper(head, null);

		private static Node ReverseLinkedListHelper(Node head, Node next)
		{
			if (head == null)
			{
				return next;
			}

			Node previous = head.Next;
			head.Next = next;

			return ReverseLinkedListHelper(previous, head);
		}

		private static void PrintLinkedList(Node head)
		{
			while (head != null)
			{
				Console.WriteLine(head.Value);
				head = head.Next;
			}
		}
	}
}